Allow axes to be centered on the chart area

etimberg opened the pull request:
Based on #2960 
Resolves #2164

Center horizontal scales draw in the center of the chart area, rather than at the bottom.

![center axis](https://user-images.githubusercontent.com/6757853/70391521-34d61e00-19a4-11ea-979b-b053f772d39a.PNG)

# To Do

- [x] Decide on the best implementation in the layout system for these boxes
- [x] Implement centerVertical mode
- [x] Sample file
- [x] Documentation
- [x] Testing

kurkle commented on src/core/core.layouts.js:
```diff
@@ -57,13 +57,14 @@ function buildLayoutBoxes(boxes) {
 	return {
 		leftAndTop: left.concat(top),
-		rightAndBottom: right.concat(bottom),
+		rightAndBottom: right.concat(bottom).concat(centerHorizontal),
```
Why in `rightAndBottom` and not in `vertical`?

etimberg replied:
`centerHorizontal` is an X axis that is vertically centered. I'm open to other names though, since I can forsee it being confusing

kurkle replied:
What about something like:
```js
scales: {
  my : {
    axis: 'x',
    position: 'center'
 }
} 
```
could also extend to `position : 'y=700' `

etimberg replied:
sure, will see how that goes

benmccann replied:
`position: 'center'` makes sense to me

etimberg replied:
Cool, I just pushed some changes. Now you specify `position: 'center'` and `axis: 'x'` in the scale options to make it work. If we like where this is going, I can look into the y axis next

kurkle replied:
The direction looks good to me!

kurkle commented on src/core/core.layouts.js:
```diff
@@ -375,7 +376,9 @@ module.exports = {
 			right: chartArea.left + chartArea.w,
-			bottom: chartArea.top + chartArea.h
+			bottom: chartArea.top + chartArea.h,
+			height: CharacterData.h,
```
`CharacterData`?

etimberg replied:
Meant to type `chartArea` and I guess autocomplete had other ideas

etimberg commented:
I added support for center Y axes too

![center both axes](https://user-images.githubusercontent.com/6757853/70395909-b7c19d80-19d1-11ea-9f06-88f801260dfb.PNG)

kurkle commented on src/core/core.controller.js:
```diff
@@ -147,8 +147,8 @@ function updateConfig(chart) {
-function positionIsHorizontal(position) {
-	return position === 'top' || position === 'bottom';
+function positionIsHorizontal(position, axis = undefined) {
+	return position === 'top' || position === 'bottom' || (position === 'center' && axis === 'x');
```
Should `axis === 'x' ` alone be enough? (future ready and simpler)

etimberg replied:
Yes, that should work. I've updated to that

kurkle commented on src/core/core.scale.js:
```diff
@@ -669,7 +669,7 @@ class Scale extends Element {
 		var gridLineOpts = opts.gridLines;
 		var display = me._isVisible();
-		var isBottom = opts.position === 'bottom';
+		var isBottom = opts.position === 'bottom' || (opts.position === 'center' && opts.axis === 'x');
```
It bothers me a bit that its considered to be bottom, when its center. Could rename the variable to something more accurate.

etimberg replied:
I think the logic was to know when labels are below the ticks, so I've renamed the variable

benmccann commented on src/core/core.controller.js:
```diff
@@ -147,8 +147,8 @@ function updateConfig(chart) {
-function positionIsHorizontal(position) {
-	return position === 'top' || position === 'bottom';
+function positionIsHorizontal(position, axis = undefined) {
+	return position === 'top' || position === 'bottom' || axis === 'x';
```
Could we get this from the `scaleId`?  I know we had talked about introducing this anyway at some point, but just as we've done elsewhere it seems like it shouldn't be necessary if the scaleId is `x`

etimberg replied:
We could. I'm ok to do that here to keep this simpler and then add an `axis` option in a separate PR.

benmccann replied:
I think it'd keep it'd make it easier for the user because then they wouldn't need to specify `axis` most of the time and it'd be more consistent that specifying `scales.x` means it's an x scale everywhere

etimberg replied:
So, I had modified it to determine the axis from the ID. That causes problems when we change the positions of an axis because the axis configs have existing defaults. I'm going to try removing the defaults entirely for axis positions, and then maybe this will be cleaned up

etimberg commented:
@kurkle I have the `position='y=..'` mode working locally. You have to specify a `positionAxisID` so that the axis knows how to look up the position using the other axis. Results for `position='y=-20'` are below

![y= positioning](https://user-images.githubusercontent.com/6757853/70398224-31b05180-19e7-11ea-96b7-bf009f804c5a.PNG)

etimberg commented:
I pushed up support for `y=...` and `x=...` modes. Includes documentation & samples. Not sure what kinds of tests are appropriate here. All the text would have to be turned off for an image test

kurkle commented:
The variable position: Maybe use object instead,
`position: {y: 500}`
It would allow `position: {x: 10, y: 50}` for radial scale.
We could require the key to be axisID, removing the need for extra option for that.
(I did not actually look at the code yet, just pondering the config format)

etimberg commented:
I'll have to look and see how that goes. It might work, but I can foresee some validation challenges. I won't get to this until tomorrow (Dec. 10) evening so I can ponder on it until then

kurkle commented on src/core/core.scale.js:
```diff
@@ -1006,12 +1006,38 @@ class Scale extends Element {
 			x2 = alignBorderValue(chartArea.right) - axisHalfWidth;
 			tx1 = borderValue + axisHalfWidth;
 			tx2 = me.left + tl;
+		} else if (axis === 'x') {
```
I think I'd rather put this logic in `layouts` and set a variable, `borderPixel` for example, to the scale (box). And use the variable here and in `_computeLabelItems`.

kurkle replied:
or just set the top/bottom correctly and use those here as if this was a bottom scale.

etimberg replied:
My initial thought was to use the layout system as well. The reason I didn't was that it would mean that the layout system would have to differentiate scales from other boxes. I wasn't sure if that was going to work long-term or not

kurkle commented on src/core/core.controller.js:
```diff
@@ -147,8 +147,8 @@ function updateConfig(chart) {
 
-function positionIsHorizontal(position) {
-	return position === 'top' || position === 'bottom';
+function positionIsHorizontal(position, axis = undefined) {
```
no need to specify `undefined` as default, it will be undefined if not provided.

kurkle commented on src/core/core.controller.js:
```diff
@@ -341,7 +341,7 @@ helpers.extend(Chart.prototype, /** @lends Chart */ {
 			var scaleType = valueOrDefault(scaleOptions.type, item.dtype);
 
-			if (positionIsHorizontal(scaleOptions.position) !== positionIsHorizontal(item.dposition)) {
+			if (positionIsHorizontal(scaleOptions.position, scaleOptions.axis) !== positionIsHorizontal(item.dposition)) {
```
could fall back to first letter of id here:
```js
scaleOptions.axis || id[0]
```

etimberg replied:
Done

kurkle commented:
Another question, does the a centered scale still reduce chartArea?

etimberg commented:
I updated this to use this format for the complex positions
```
position: {
  y: 20
}
```

kurkle commented:
Some lint errors, but looks good!

etimberg commented:
Sweet, I fixed the lint warnings

etimberg commented:
Will look into the test failures. Not sure what I broke.

etimberg commented:
I fixed the test failures. The cause was defaulting the `axis` option based on `id[0]`. Because if its` x` and the axis has a default position of `left` the code doesn't reassign the scale to the bottom.

For now, I removed the `id[0]` default. We could try removing the axis position defaults from the scale defaults and rely on the default position setting

etimberg commented:
I've added fixture based tests for this so its good to review

benmccann commented:
Looks good to me. I still think it'd be nice if `axis` weren't required, but otherwise am happy with it

etimberg commented:
@benmccann I removed the axis defaults, and handled it correctly. Now we can use the first character of the axis ID to determine if its horizontal or vertical

benmccann commented on docs/axes/cartesian/README.md:
```diff
@@ -14,13 +14,30 @@ All of the included cartesian axes support a number of common options.
 | `type` | `string` | | Type of scale being employed. Custom scales can be created and registered with a string key. This allows changing the type of an axis for a chart.
-| `position` | `string` | | Position of the axis in the chart. Possible values are: `'top'`, `'left'`, `'bottom'`, `'right'`
+| `position` | `string` | | Position of the axis. [more...](#axis-position)
+| `axis` | `string` | | Which type of axis this is. Possible values are: `'x'`, `'y'`.
```
Maybe add something here about how it will be automatically determined from first letter of id if not specified

etimberg replied:
Done

benmccann commented on docs/axes/cartesian/README.md:
```diff
@@ -14,13 +14,30 @@ All of the included cartesian axes support a number of common options.
+An axis can either be positioned at the edge of the chart, at the center of the chart area, or dynamically with respect to a data value.
+
+To position the axis at the edge of the chart, set the `position` option to one of: `'top'`, `'left'`, `'bottom'`, `'right'`.
+To position the axis at the center of the chart area, set the `position` option to `'center'`. In this mode, the `axis` option must be specified.
```
"In this mode, the `axis` option must be specified." - probably not true anymore

etimberg replied:
Updated

benmccann commented on src/core/core.controller.js:
```diff
@@ -147,8 +147,9 @@ function updateConfig(chart) {
 
-function positionIsHorizontal(position) {
-	return position === 'top' || position === 'bottom';
+const WELL_KNOWN_POSITIONS = ['top', 'bottom', 'left', 'right', 'chartArea'];
```
I might call this just `KNOWN_POSITIONS`

etimberg replied:
Done

benmccann commented on src/core/core.controller.js:
```diff
@@ -147,8 +147,9 @@ function updateConfig(chart) {
-function positionIsHorizontal(position) {
-	return position === 'top' || position === 'bottom';
+const WELL_KNOWN_POSITIONS = ['top', 'bottom', 'left', 'right', 'chartArea'];
+function positionIsHorizontal(position, axis) {
```
can we just do `axis === 'x'` instead?

etimberg replied:
We might now that I removed default positions. Before I did that, it would override the fact that the position was wrong and the config would not be updated.

etimberg replied:
Tried this. It breaks the default position fixing because it overrides the wrong axis position value. Lets leave as is for now
